src/main.rs

```diff
@@ -38,13 +54,55 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let addr = "0.0.0.0:4243".parse().unwrap();
     tracing::info!("chariott listening on {addr}");
 
-    let router = Server::builder()
-        .add_service(ChariottServiceServer::new(ChariottServer::new(registry, broker)));
+    let server = Arc::new(ChariottServer::new(registry, broker));
+    let router =
+        Server::builder().add_service(ChariottServiceServer::from_arc(Arc::clone(&server)));
 
     #[cfg(build = "debug")]
     let router = router.add_service(reflection_service);
 
-    router.serve_with_ctrl_c_shutdown(addr).await?;
+    let error_cancellation_token = CancellationToken::new();
+    let ctrl_c_cancellation_token = ctrl_c_cancellation();
+
+    let prune_loop = {
+        use tokio::{select, spawn, time::sleep};
+
+        let ctrl_c_cancellation_token = ctrl_c_cancellation_token.clone();
+        let error_cancellation_token = error_cancellation_token.child_token();
+
+        spawn(async move {
```

Comment:
Why do we spawn this? We could just return it as a `Future` and then when we call `tokio::join!` it will start to make progress?

Reply:
This was a refactoring oversight. I made `router_serve` a future, but forgot to come back to this one. Thanks for spotting. Addressed with 2bd778b30676bb984efc8d9c151fe3e5b66c85d6.